cmd/proxy/wsl_integration_linux.go

```diff
@@ -48,11 +52,25 @@ func main() {
 		logrus.Fatalf("invalid upstream URL: %s", upstreamAddr)
 	}
 
+	ctx, _ := signal.NotifyContext(context.Background(), syscall.SIGINT, syscall.SIGTERM)
 	proxy := httputil.NewSingleHostReverseProxy(targetURL)
 
+	srv := http.Server{
+		Addr:              listenAddr,
+		Handler:           proxy,
+		ReadHeaderTimeout: 5 * time.Second,
+	}
+
+	go func() {
+		if err := srv.ListenAndServe(); err != nil {
+			logrus.Error("Error starting server:", err)
+		}
+	}()
+
 	logrus.Debugf("proxy server is running on %s", listenAddr)
-	err = http.ListenAndServe(listenAddr, proxy) //nolint:gosec // No security concern for not having timeout here.
-	if err != nil {
-		logrus.Error("Error starting server:", err)
+	<-ctx.Done()
+
+	if err := srv.Shutdown(ctx); err != nil {
```

Comment:
Calling `Shutdown` with a context that's already done is an error; this means that we won't give any time for the actual guts of `Shutdown`.

Instead of putting `ListenAndServe` in a goroutine, it might be better to do it the other way around:
```go
go func() {
  <-ctx.Done()
  if err := srv.Shutdown(context.Background()); err != nil {
    logrus.Error(…
```
(`ListenAndServe` will return [`ErrServerClosed`](https://pkg.go.dev/net/http@go1.22.0#ErrServerClosed) either way, so you probably don't want to log if that's the returned error.)

Reply:
Ah, yes. That was meant to be `context.Background()`